contracts/src/v0.8/dev/vrf/TrustedBlockhashStore.sol

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.6;
+
+import "../../ChainSpecificUtil.sol";
+import "../../shared/access/ConfirmedOwner.sol";
+import "./BlockhashStore.sol";
+
+contract TrustedBlockhashStore is ConfirmedOwner, BlockhashStore {
+  error NotInWhitelist();
+  error InvalidTrustedBlockhashes();
+  error InvalidRecentBlockhash();
+
+  mapping(address => bool) public s_whitelistStatus;
+  address[] public s_whitelist;
+
+  constructor(address[] memory whitelist) ConfirmedOwner(msg.sender) {
+    setWhitelist(whitelist);
+  }
+
+  /**
+   * @notice sets the whitelist of addresses that can store blockhashes
+   * @param whitelist the whitelist of addresses that can store blockhashes
+   */
+  function setWhitelist(address[] memory whitelist) public onlyOwner {
+    address[] memory previousWhitelist = s_whitelist;
+    s_whitelist = whitelist;
+
+    // Unset whitelist status for all addresses in the previous whitelist,
+    // and set whitelist status for all addresses in the new whitelist.
+    for (uint256 i = 0; i < previousWhitelist.length; i++) {
+      s_whitelistStatus[previousWhitelist[i]] = false;
+    }
+    for (uint256 i = 0; i < whitelist.length; i++) {
+      s_whitelistStatus[whitelist[i]] = true;
+    }
+  }
+
+  /**
+   * @notice stores a list of blockhashes and their respective blocks, only callable
+   * by a whitelisted address
+   * @param blockhashes the list of blockhashes and their respective blocks
+   */
+  function storeTrusted(
+    uint256[] calldata blockNums,
+    bytes32[] calldata blockhashes,
+    uint256 recentBlockNumber,
+    bytes32 recentBlockhash
+  ) external {
+    bytes32 onChainHash = ChainSpecificUtil.getBlockhash(uint64(recentBlockNumber));
+    if (onChainHash != recentBlockhash) {
+      revert InvalidRecentBlockhash();
+    }
```

Comment:
This protects against re-orgs, though will we successfully re-queue a tx with an updated `recentBlockhash` argument in that case?

Reply:
Yes. The off-chain loop runs `GetUnfulfilledBlocksAndRequests` which gets the present state of the coordinator. If the blockhash store reverts due to a re-org those dependent requests will once again be noticed as unfulfilled.

Reply:
The exception is if for some reason those requests fell out of the lookback window between the re-org and them being noticed again, which is unlikely due to the fact that the lookback window can be quite generous.

Reply:
so is it safe to say trusted blockhash store can be used for blockchains with re-orgs? It seems so because any fulfillments that happen against a re-orged blockhash will also be re-orged away and require re-fulfillments with canonical blockhash?